File: OneQuestionPerDay/238/product_of_array_except_self.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [0] * len(nums)
        k = 1
        for i in range(len(nums)):
            res[i] = k
            k *= nums[i]
        k = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= k
            k *= nums[i]
        return res

File: OneQuestionPerDay/238/product_of_array_except_self.go
package main

func productExceptSelf(nums []int) []int {
	res := make([]int, len(nums))
	res[0] = 1
	for i := 1; i < len(nums); i++ {
		res[i] = res[i-1] * nums[i-1]
	}
	r := 1
	for i := len(nums)-2;i >= 0;i-- {
		r *= nums[i+1]
		res[i] *= r
	}
	return res
}
